packages/core/src/emitter/index.ts

```diff
@@ -1,29 +1,61 @@
-export class Emitter<EventName extends string = string> {
-  private callbacks: Partial<Record<EventName, Function[]>> = {}
+type EventName = string
+type EventFnArgs = any[]
+type EmitterContract = Record<EventName, EventFnArgs>
 
-  on(event: EventName, callback: Function): this {
-    this.callbacks[event] = [...(this.callbacks[event] ?? []), callback]
+/**
+ * Event Emitter that takes the expected contract as a generic
+ * @example
+ * ```ts
+ *  type Contract = {
+ *    delivery_success: [DeliverySuccessResponse, Metrics],
+ *    delivery_failure: [DeliveryError]
+ * }
+ *  new Emitter<Contract>()
+ *  .on('delivery_success', (res, metrics) => ...)
+ *  .on('delivery_failure', (err) => ...)
```

Comment:
Love that you added an example here - it gets hard to parse the types sometimes :smile:

Reply:
Thanks!